viostor/virtio_pci.c

```diff
@@ -197,12 +197,12 @@ static u16 vdev_get_msix_vector(void *context, int queue)
             } else {
                 vector = queue + 1;
             }
-        } else {
-            vector = VIRTIO_MSI_NO_VECTOR;
         }
     } else {
         /* on-device-config-change interrupt */
-        vector = VIRTIO_MSI_NO_VECTOR;
+        if (!adaptExt->msix_one_vector) {
```

Comment:
My preferred solution is to fix the code to be fully functional with any number of vectors that the QEMU device supports. Look at what Linux virtio driver do in vp_find_vqs. They first try to allocate one vector per queue + one for config. Failing that, they try one for all queues and one for config (2 total). And if even that doesn't work they fall back to IRQ. Not sure if we can do the IRQ fallback on Windows but at least the fallback to 2 vectors is doable.

With the code proposed in this patch you're going to disable config interrupts if the number of vectors is less than number of queues + 1. That's not necessary.

This condition in VirtIoHwInitialize should be refined:

```
    if (adaptExt->msix_vectors >= (adaptExt->num_queues + 1U)) {
        /* initialize queues with a MSI vector per queue */
        RhelDbgPrint(TRACE_LEVEL_INFORMATION, ("Using a unique MSI vector per queue\n"));
        adaptExt->msix_one_vector = FALSE;
    } else {
        /* if we don't have enough vectors, use one for all queues */
        RhelDbgPrint(TRACE_LEVEL_INFORMATION, ("Using one MSI vector for all queues\n"));
        adaptExt->msix_one_vector = TRUE;
    }

```
To distinguish between "at least 2 but less than num_queues + 1" and "1" and disable config interrupts only in the latter case.

Also, looking at the code you're changing in VirtIoMSInterruptRoutine, how are we supposed to handle all queues with one vector? You just assign 1 to MessageID. There needs to be a loop somewhere. See how VioScsiMSInterrupt does it.

I agree that trying to handle everything with one vector wouldn't work well with viostor. We can't even use the ISR status register in MSI mode so we would have to read the entire configuration on each interrupt to see if anything changed - that's way too suboptimal, even for "limp mode". Running with no config interrupts if there is no other choice is reasonable. We could poll the config periodically but I'd leave that to the over-engineering department :)

If you believe that reporting one vector is a HW bug, then we should submit a QEMU patch to fix it.